### PR TITLE
RuleMatcher: Use a single list for tracking pattern_matches

### PR DESCRIPTION
**Pull request by awelzel**
@rsmmr - as discussed on #3996. How does that look, other thoughts?

### CODE REVIEW

File: src/RuleMatcher.cc [hunk 1]
```diff
@@ -865,11 +875,10 @@ void RuleMatcher::Match(RuleEndpointState* state, Rule::PatternType type, const
                 continue;
 
             // Remember that all patterns have matched.
-            if ( ! state->matched_by_patterns.is_member(r) ) {
-                state->matched_by_patterns.push_back(r);
-                String* s = new String(data, data_len, false);
-                state->matched_text.push_back(s);
-                state->matched_text_end_of_match.push_back(match_end_pos - pre_match_pos);
+            const auto* pattern_match = state->FindRulePatternMatch(r);
+            if ( pattern_match == nullptr ) {
```
Comment (rsmmr) [inline src/RuleMatcher.cc:879]: Style nit, but I think we generally skip explicit comparisons to `nullptr`:

```suggestion
            if ( ! pattern_match ) {
```

File: src/RuleMatcher.cc [hunk 2]
```diff
@@ -893,8 +902,8 @@ void RuleMatcher::FinishEndpoint(RuleEndpointState* state) {
 
     ExecPureRules(state, true);
 
-    loop_over_list(state->matched_by_patterns, i)
-        ExecRulePurely(state->matched_by_patterns[i], state->matched_text[i], state, true);
+    std::for_each(state->pattern_matches.begin(), state->pattern_matches.end(),
+                  [this, state](auto& m) { ExecRulePurely(m.rule, m.text, state, true); });
```
Comment (rsmmr) [inline src/RuleMatcher.cc:906]: This is my personal preference, so YMMV, but I do prefer a simple for-loop over `std::for_each` any time ...

File: src/RuleMatcher.cc [hunk 3]
```diff
@@ -997,11 +1006,11 @@ void RuleMatcher::ExecRule(Rule* rule, RuleEndpointState* state, bool eos) {
                 return;
             }
 
+
```
Comment (rsmmr) [inline src/RuleMatcher.cc:1009]:
```suggestion
```

File: src/RuleMatcher.h [hunk 4]
```diff
@@ -183,13 +213,8 @@ class RuleEndpointState {
...
     rule_hdr_test_list hdr_tests;
 
-    // The follow tracks which rules for which all patterns have matched,
-    // in a parallel list the (first instance of the) corresponding
-    // matched text, and in another parallel list the offset of the
-    // end of the last pattern match.
-    rule_list matched_by_patterns;
-    bstr_list matched_text;
-    match_offset_list matched_text_end_of_match;
+    // The follow all pattern matches for rules for which all patterns have matched.
```
Comment (rsmmr) [inline src/RuleMatcher.h:216]:
```suggestion
    // The following tracks all pattern matches for rules for which all patterns have matched.
```

File: src/RuleMatcher.cc (hunk 1 above)
Comment (rsmmr) [reply src/RuleMatcher.cc:879]: Actually, don't even need the tmp:

```
if ( ! state->FindRulePatternMatch(r) ) {
```

File: src/RuleMatcher.h [hunk 5]
```diff
@@ -168,6 +169,35 @@ class RuleEndpointState {
     RuleEndpointState(analyzer::Analyzer* arg_analyzer, bool arg_is_orig, RuleEndpointState* arg_opposite,
                       analyzer::pia::PIA* arg_PIA);
 
+    // Tracking pattern matches for a given Rule.
+    class RulePatternMatch {
```
Comment (rsmmr) [inline src/RuleMatcher.h:173]: Optional: if all members are public, I personally like making it a `struct` to indicate so.

File: src/RuleMatcher.h [hunk 6]
```diff
@@ -168,6 +169,35 @@ class RuleEndpointState {
...
+    public:
+        RulePatternMatch(Rule* rule, const u_char* data, int data_len, MatchPos end_of_match)
+            : rule(rule), text(new String(data, data_len, false)), end_of_match(end_of_match) {}
```
Comment (rsmmr) [inline src/RuleMatcher.h:176]: Suggest to use `unique_ptr` for `text` to avoid manual memory mgmt.

Comment (rsmmr) [reply src/RuleMatcher.h:176]: Or, actually, do we need this on the heap at all, vs. just storing a `String` directly inside the `class`?

**Review by rsmmr — APPROVED** [review]
Look good, just small notes.

File: src/RuleMatcher.cc (hunk 2 above)
Comment (awelzel) [reply src/RuleMatcher.cc:906]: Yep, thanks for mentioning!

**Pull request merged**